presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java

```diff
@@ -68,6 +80,22 @@ public HudiDirectoryLister(Configuration conf, ConnectorSession session, Table t
                 .setConf(actualConfig)
                 .setBasePath(table.getStorage().getLocation())
                 .build();
+        String latestCompactionInstantTime = metaClient.getActiveTimeline()
+                .getCommitsAndCompactionTimeline()
+                .filterCompletedInstants()
+                .filter(instant -> HoodieTableType.MERGE_ON_READ.equals(metaClient.getTableType()) && instant.getAction().equals(HoodieTimeline.COMPACTION_ACTION))
+                .lastInstant()
+                .map(HoodieInstant::getTimestamp).orElse(null);
+        if (latestCompactionInstantTime != null) {
+            this.latestInstant = latestCompactionInstantTime;
+        }
+        else {
+            this.latestInstant = metaClient.getActiveTimeline()
+                    .getCommitsTimeline()
+                    .filterCompletedInstants()
+                    .lastInstant()
+                    .map(HoodieInstant::getTimestamp).orElseThrow(() -> new RuntimeException("No active instant found"));
+        }
```

Comment:
I feel like we should go with the fully functional approach rather than mixing imperative styles here since you're already using the optional type.

```suggestion
this.latestInstant = metaClient.getActiveTimeline()
                .getCommitsAndCompactionTimeline()
                .filterCompletedInstants()
                .filter(instant -> HoodieTableType.MERGE_ON_READ.equals(metaClient.getTableType()) && instant.getAction().equals(HoodieTimeline.COMPACTION_ACTION))
                .lastInstant()
                .map(HoodieInstant::getTimestamp).orElseGet(() -> metaClient.getActiveTimeline()
                        .getCommitsTimeline()
                        .filterCompletedInstants()
                        .lastInstant()
                        .map(HoodieInstant::getTimestamp).orElseThrow(() -> new RuntimeException("No active instant found")));
```